### ISSUE TITLE
False positive when calling `shutil.which` with a `str`

### ISSUE THREAD

**Issue by xavdid** [microsoft/pyright#5234]
**Describe the bug**

Pyright raises a type error when calling `shutil.which` with a string.

**To Reproduce**

Call `shutil.which` with a string

**Expected behavior**

No type error

**Screenshots or Code**

```python
import shutil

if "a" in [] or shutil.which("pytest"):
    pass
```

(_edit_: updated snippet)

gives:

```
Argument of type "Literal['pytest']" cannot be assigned to parameter "cmd" of type "_StrPathT@which" in function "which"
  Type "Literal['pytest']" cannot be assigned to type "bool"
    "Literal['pytest']" is incompatible with "bool"
Argument of type "Literal['pytest']" cannot be assigned to parameter "cmd" of type "bytes" in function "which"
  "Literal['pytest']" is incompatible with "bytes"
```

**VS Code extension or command-line**

Running pyright via Pylance `v2023.5.50`. 

**Additional context**

- typeshed defines `which` to take a `_StrPathT` ([python/typeshed | shutil.pyi#L166](https://github.com/python/typeshed/blob/a7702164349776ebf25e2c1e0a0c123333a5b68f/stdlib/shutil.pyi#L166))
- this is bound to `StrPath` ([python/typeshed | shutil.pyi#L38](https://github.com/python/typeshed/blob/a7702164349776ebf25e2c1e0a0c123333a5b68f/stdlib/shutil.pyi#L38))
- which is defined as `str | PathLike[str]` ([python/typeshed | __init__.pyi#L130](https://github.com/python/typeshed/blob/a7702164349776ebf25e2c1e0a0c123333a5b68f/stdlib/_typeshed/__init__.pyi#L130))

Also, it works as expected on mypy ([playground](https://mypy-play.net/?mypy=latest&python=3.11&gist=2353a5c8fca7d9bb80a88e69e9cfb4cd))

**Comment by erictraut** [comment]
I'm not able to repro the problem with the code sample provided above. Can you confirm that you are able to repro the error with this code sample pasted into a new file? If so, can you still repro it with a freshly-created venv? I'm wondering if you have a stale stub library installed in your venv.

If you right click on `where` and choose "Go to Declaration", where does it take you? If you don't have a stale stub library installed, it should take you to `extensions/ms-python.vscode-pylance-2023.5.50/dist/typeshed-fallback/stdlib/shutil.pyi`.

**Comment by xavdid** [comment]
Apologies - seems like the error only shows when it's the second part of a conditional. Here's an updated snippet which shows the error:

```py
import shutil

if "a" in [] or shutil.which("pytest"):
    pass
```

I've confirmed "Go to Declaration" matches the path you posted. 

---

I saw the type error in my code and didn't realize the conditional was required to get the bug to show; i'll test more carefully in the future 🙈

**Comment by erictraut** [comment]
Thanks for the updated code snippet. I'm able to repro the problem now.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.312, which I just published. It will also be included in the next insiders release of pylance.
